### PR TITLE
add GLEM model, TAGDataset and example of GLEM

### PR DESCRIPTION
**Pull request by ECMGit**
reopened  #9591 

Feature summary:

- Add GLEM as GNN & LLM Co-training model to PyG
- adapt GLEM's LM to AutoModelForSequenceClassification from transformers
- Lora support
- LM/LLM support
- ogbn-products/ogbn-arxiv testing finished
- TAGDataset can be used as a wrapper class for any node classification dataset in PyG with LM tokenizer and associate raw text
- external prediction as pseudo labels supported

### CODE REVIEW

**Review by puririshi98 — APPROVED** [review]
LGTM just get CI green

**Comment by puririshi98** [comment]
@rusty1s @akihironitta ready for your reviews

File: examples/llm/glem.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+import torch
+
+from torch_geometric import seed_everything
+from torch_geometric.data import download_google_url
+from torch_geometric.datasets import TAGDataset
+from torch_geometric.nn.models import GAT, GCN, GLEM, GraphSAGE
+
+# Add the parent directory to sys.path
+parent_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.append(parent_dir)
```
Comment (akihironitta) [inline examples/llm/glem.py:30]: Why is this necessary?

File: examples/llm/README.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
...
+| Example                              | Description                                                                                                                                                                                    |
+| ------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [`g_retriever.py`](./g_retriever.py) | Example for Retrieval-Augmented Generation (RAG) w/ GNN+LLM by co-training `LLAMA2` with `GAT` for answering questions based on knowledge graph information                                    |
+| [`glem.py`](./glem.py)               | Example for [GLEM](https://arxiv.org/abs/2210.14709), a GNN+LLM co-training model via variational Expectation-Maximization (EM) framework on node classification tasks to achieve SOTA results |
+
+## Run GLEM for getting SOTA result on ogbn-products dataset
+
+`python glem.py`
```
Comment (akihironitta) [inline examples/llm/README.md:10]:
```suggestion
```

File: examples/llm/glem.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    device = torch.device(
+        f'cuda:{gpu}' if torch.cuda.is_available() else 'cpu')
+    print(f'Running on: {torch.cuda.get_device_name({gpu})}')
+    torch.cuda.empty_cache()
+
+    if not train_without_ext_pred:
+        ext_pred_path = download_google_url(
+            id='15sO2m7BeW7C1Upmdw3Cx1JS__6nxTAzY',
+            folder='/work/users/junhaos/glem_data/ogbn_products/ext_preds',
+            filename='giant_sagn_scr.pt', log=True)
```
Comment (akihironitta) [inline examples/llm/glem.py:76]: Let's use a relative path for other people to use.

File: examples/llm/glem.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        ext_pred = torch.load(ext_pred_path, map_location=device)
+        ext_pseudo_labels = ext_pred.argmax(dim=-1)
+        pretrain_augmented = True
+
+    seed_everything(42)
+    from ogb.nodeproppred import PygNodePropPredDataset
```
Comment (akihironitta) [inline examples/llm/glem.py:82]: nit: Let's move the import statement at the start of the file.

File: examples/llm/glem.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    print(f'Total running time: {running_time:.2f} hours')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='GLEM Example:')
+    parser.add_argument('--gpu', type=int, default=0)
+    parser.add_argument('--num_runs', type=int, default=10,
+                        help='number of runs')
+    parser.add_argument('--num_em_iters', type=int, default=1,
+                        help='number of iterations')
```
Comment (akihironitta) [inline examples/llm/glem.py:390]: Is `1` a good default value for EM iterations?

File: examples/llm/glem.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        if em_phase == 'gnn':
+            gnn_test_acc = max(gnn_test_acc, final_test_acc)
+            model.gnn = model.gnn.to('cpu', non_blocking=True)
+            em_phase = 'lm'
+        else:
+            lm_test_acc = max(lm_test_acc, final_test_acc)
+            model.lm = model.lm.to('cpu', non_blocking=True)
+            em_phase = 'gnn'
+        torch.cuda.empty_cache()
+    print(f'Best GNN acc: {gnn_test_acc}, LM acc: {lm_test_acc}')
```
Comment (akihironitta) [inline examples/llm/glem.py:377]: This is the same comment as https://github.com/pyg-team/pytorch_geometric/pull/9467#discussion_r1786532509, but we shouldn't pick the best metric evaluated on the test set at the end of every EM step.

**Review by akihironitta — COMMENTED** [review]
Could we have type annotations all over the PR? Also, I'd suggest splitting this PR into smaller ones.

**Review by akihironitta — COMMENTED** [review]
I haven't had a look outside the example script yet, but this addition is exciting! 🚀

File: examples/llm/glem.py (hunk 5 above)
Comment (ECMGit) [reply examples/llm/glem.py:390]: Yes, I have tested it till 5, mostly 1-2 EM iteration can perform very well on ogbn-products and ogbn-arxiv

File: examples/llm/glem.py (hunk 4 above)
Comment (ECMGit) [reply examples/llm/glem.py:82]: done, moved to the start of the file

File: examples/llm/glem.py (hunk 3 above)
Comment (ECMGit) [reply examples/llm/glem.py:76]: done

File: examples/llm/README.md (hunk 2 above)
Comment (ECMGit) [reply examples/llm/README.md:10]: done

File: examples/llm/glem.py (hunk 6 above)
Comment (ECMGit) [reply examples/llm/glem.py:377]: Hi Akihiro,

Thanks for reviewing the code.

I think the case is different here, I agree that we should not pick best test metrics after every epoch, but the test metrics still required for every EM step. Since E step is LM training and M-step is GNN training, both step need certain number of epochs. We need to run full inference after every E and M step for finding out that which model have better result.

File: examples/llm/glem.py (hunk 1 above)
Comment (ECMGit) [reply examples/llm/glem.py:30]: Removed, originally I used it for unifying data folder for other script.

File: examples/llm/glem.py (hunk 6 above)
Comment (puririshi98) [reply examples/llm/glem.py]: @ECMGit i think @akihironitta's point is that to "finding out that which model have better result" you should only use val accuracy, not test accuracy since if you use the test acc this could be viewed as a form of loosely fitting the model to the test set

Comment (ECMGit) [reply examples/llm/glem.py]: done

**Comment by puririshi98** [comment]
LGTM @akihironitta @rusty1s  let us know if anything else needed

**Comment by puririshi98** [comment]
merging this along with community sprint PRs since matthias is busy till end of year and he told me to test/review/merge them myself
